### PR TITLE
TC06 TC07 TC09

### PR DESCRIPTION
**Pull request by viethuynh140499**
TC06 - **Additional pages display once user logged in**
TC07 - **User can create new account**
TC09 - **User can change password**

### CODE REVIEW

File: src/test/java/test_cases/ChangePasswordTest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    loginPage.login(user);
+    homePage.clickChangePasswordTab();
+
+    user.setPassword(DataHelper.getRandomValidPassword());
+    String newPassword = user.getPassword();
+
+    changePasswordPage.changePassword(oldPassword, newPassword, newPassword);
+    String actualSuccessMessage = changePasswordPage.getSuccessfullyMessage();
+    String expectedSuccessMessage = "Your password has been updated!";
+    Assert.assertEquals(actualSuccessMessage, expectedSuccessMessage, "Change Password message not match");
```
Comment (fey031) [inline src/test/java/test_cases/ChangePasswordTest.java:38]: NewPassword and ConfirmedPassword are the same, so the assert message should not be `Change Password message not match`

File: src/main/java/page_objects/ChangePasswordPage.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package page_objects;
+
+import common.element.Button;
+import common.element.Label;
+import common.element.TextBox;
+import model.User;
+import org.openqa.selenium.By;
+
+public class ChangePasswordPage extends GeneralPage{
+  private final TextBox txtCurrentPassword = new TextBox(By.cssSelector("#currentPassword"));
```
Comment (thanhlelgg) [inline src/main/java/page_objects/ChangePasswordPage.java:10]: use by Id

File: src/main/java/page_objects/ChangePasswordPage.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  private final TextBox txtNewPassword = new TextBox(By.cssSelector("#newPassword"));
+  private final TextBox txtConfirmPassword = new TextBox(By.cssSelector("#confirmPassword"));
+  private final Button btnChangePassword = new Button(By.cssSelector("[title='Change password']"));
+  private final Label lblChangePassword = new Label(By.cssSelector(".message"));
```
Comment (thanhlelgg) [inline src/main/java/page_objects/ChangePasswordPage.java:14]: use by class

File: src/test/java/test_cases/LoginTests.java [hunk 4]
```diff
@@ -48,15 +47,27 @@ public void TC03() {
...
+
+  @Test(description = "Additional pages display once user logged in")
+  public void TC06() {
+    User user = new User(Constant.USERNAME, Constant.PASSWORD);
+    homePage.clickLoginTab();
+    loginPage.login(user);
+
+    Assert.assertTrue(homePage.isMyTicketTabDisplayed(), "My ticket Tab does not display");
+    Assert.assertTrue(homePage.isLogOutTabDisplayed(), "Log out Tab does not display");
+    Assert.assertTrue(homePage.isChangePasswordTabDisplayed(), "Change Password does not display");
```
Comment (fey031) [inline src/test/java/test_cases/LoginTests.java:71]: Missing checkpoints:
click tab and verify page is displayed

File: src/main/java/page_objects/ChangePasswordPage.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  private final Label lblConfirmPasswordErrorMessage = new Label(By.cssSelector(".confirm-password label:last-child"));
```
Comment (thanhlelgg) [inline src/main/java/page_objects/ChangePasswordPage.java:15]: this xpath is not good, don't use position to capture xpath if there's a better way

**Review by fey031 — CHANGES_REQUESTED** [review]

File: src/main/java/page_objects/GeneralPage.java [hunk 6]
```diff
@@ -6,23 +6,42 @@
 
 public class GeneralPage {
     private final Tab tabLogin = new Tab(By.cssSelector("li a[href$='Login.cshtml']"));
-    private final Label lblWelcomeMessage = new Label(By.cssSelector(".account"));
+    private final Tab tabRegister = new Tab(By.cssSelector("li a[href$='Register.cshtml']"));
     private final Tab tabBookTicket = new Tab(By.cssSelector("li a[href$='BookTicketPage.cshtml'"));
+    private final Tab tabMyTicket = new Tab(By.cssSelector("li a[href$='ManageTicket.cshtml']"));
+    private final Tab tabChangePassword = new Tab(By.cssSelector("li a[href$='ChangePassword.cshtml']"));
+    private final Tab tabLogout = new Tab(By.cssSelector("li a[href*=Logout]"));
+    private final Label lblWelcomeMessage = new Label(By.cssSelector(".account"));
```
Comment (thanhlelgg) [inline src/main/java/page_objects/GeneralPage.java:14]: use by class

File: src/main/java/page_objects/RegisterPage.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import org.openqa.selenium.By;
+
+public class RegisterPage extends GeneralPage{
+  private final TextBox txtEmail = new TextBox(By.id("email"));
+  private final TextBox txtPassword = new TextBox(By.id("password"));
+  private final TextBox txtConfirmPassword = new TextBox(By.id("confirmPassword"));
+  private final TextBox txtPid = new TextBox(By.id("pid"));
+
+  private final Button btnRegister = new Button(By.cssSelector(".form-actions input"));
+  private final Label lblSuccessfully = new Label(By.cssSelector("#content"));
```
Comment (thanhlelgg) [inline src/main/java/page_objects/RegisterPage.java:16]: use by id

File: src/main/java/page_objects/RegisterPage.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+import model.User;
+import org.openqa.selenium.By;
+
+public class RegisterPage extends GeneralPage{
+  private final TextBox txtEmail = new TextBox(By.id("email"));
+  private final TextBox txtPassword = new TextBox(By.id("password"));
+  private final TextBox txtConfirmPassword = new TextBox(By.id("confirmPassword"));
+  private final TextBox txtPid = new TextBox(By.id("pid"));
+
+  private final Button btnRegister = new Button(By.cssSelector(".form-actions input"));
```
Comment (thanhlelgg) [inline src/main/java/page_objects/RegisterPage.java:15]: this xpath is pretty weak, use something more specific and descriptive

File: src/test/java/test_cases/ChangePasswordTest.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  HomePage homePage = new HomePage();
+  RegisterPage registerPage = new RegisterPage();
+  LoginPage loginPage = new LoginPage();
+  ChangePasswordPage changePasswordPage = new ChangePasswordPage();
+
+  @Test(description = "User can change password")
+  public void TC09() {
+    User user = new User();
+    user.setUsername(DataHelper.getRandomValidEmail());
+    user.setPassword(DataHelper.getRandomValidPassword());
```
Comment (thanhlelgg) [inline src/test/java/test_cases/ChangePasswordTest.java:22]: should be like this:
```
String oldPassword = DataHelper.getRandomValidPassword();
user.setPassword(oldPassword );
```

File: src/test/java/test_cases/ChangePasswordTest.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    String oldPassword = user.getPassword();
+    user.setPid(DataHelper.getRandomValidPID());
+
+    homePage.clickRegisterTab();
+    registerPage.register(user);
+    homePage.clickLoginTab();
+    loginPage.login(user);
+    homePage.clickChangePasswordTab();
+
+    user.setPassword(DataHelper.getRandomValidPassword());
```
Comment (thanhlelgg) [inline src/test/java/test_cases/ChangePasswordTest.java:32]: like above

File: src/main/java/page_objects/RegisterPage.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
+package page_objects;
+
+import common.element.Button;
+import common.element.Label;
+import common.element.TextBox;
+import model.User;
+import org.openqa.selenium.By;
+
+public class RegisterPage extends GeneralPage{
```
[the remark this answers is not in the document]
Comment (viethuynh140499) [reply src/main/java/page_objects/RegisterPage.java:9]: I named it by Tab Name in UI.

File: src/test/java/test_cases/ChangePasswordTest.java (hunk 1 above)
Comment (thanhlelgg) [inline src/test/java/test_cases/ChangePasswordTest.java:38]: fail message isn't clear and there's grammar mistake. you should just use "Change password unsuccessfully" or sth like that

File: src/main/java/page_objects/RegisterPage.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  private final TextBox txtPid = new TextBox(By.id("pid"));
+
+  private final Button btnRegister = new Button(By.cssSelector(".form-actions input"));
+  private final Label lblSuccessfully = new Label(By.cssSelector("#content"));
+
+  public String getSuccessfullyMessage() {
+    return this.lblSuccessfully.getText();
+  }
+
+  public void register(User user) {
```
[the remark this answers is not in the document]
Comment (viethuynh140499) [reply src/main/java/page_objects/RegisterPage.java:22]: I changed method name to registerAccount

File: src/test/java/test_cases/RegisterTests.java [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    user.setUsername(DataHelper.getRandomValidEmail());
+    user.setPassword(DataHelper.getRandomValidPassword());
+    user.setPid(DataHelper.getRandomValidPID());
+
+    homePage.clickRegisterTab();
+    registerPage.register(user);
+
+    String actualSuccessMessage = registerPage.getSuccessfullyMessage();
+    String expectedSuccessMessage = "You're here";
+    Assert.assertEquals(actualSuccessMessage,expectedSuccessMessage, "Success message does not display");
```
Comment (thanhlelgg) [inline src/test/java/test_cases/RegisterTests.java:26]: spacing

**Review by thanhlelgg — CHANGES_REQUESTED** [review]

File: src/test/java/test_cases/LoginTests.java (hunk 4 above)
Comment (viethuynh140499) [reply src/test/java/test_cases/LoginTests.java:71]: Yes, I added checkpoints click tab and verify page is displayed

File: src/main/java/page_objects/ChangePasswordPage.java (hunk 2 above)
Comment (viethuynh140499) [reply src/main/java/page_objects/ChangePasswordPage.java:10]: I edited into By.id

File: src/test/java/test_cases/ChangePasswordTest.java (hunk 1 above)
Comment (viethuynh140499) [reply src/test/java/test_cases/ChangePasswordTest.java:38]: yes, i edited into Change password unsuccessfully

File: src/test/java/test_cases/ChangePasswordTest.java (hunk 10 above)
Comment (viethuynh140499) [reply src/test/java/test_cases/ChangePasswordTest.java:32]: I fixed it. Thank you.

File: src/main/java/page_objects/ChangePasswordPage.java (hunk 3 above)
Comment (viethuynh140499) [reply src/main/java/page_objects/ChangePasswordPage.java]: I change into By.class

File: src/main/java/page_objects/ChangePasswordPage.java (hunk 5 above)
Comment (viethuynh140499) [reply src/main/java/page_objects/ChangePasswordPage.java:15]: I fixed it. Thank you.

File: src/main/java/page_objects/RegisterPage.java (hunk 8 above)
Comment (viethuynh140499) [reply src/main/java/page_objects/RegisterPage.java]: I fixed it. Thank you.

File: src/test/java/test_cases/ChangePasswordTest.java (hunk 9 above)
Comment (viethuynh140499) [reply src/test/java/test_cases/ChangePasswordTest.java]: I fixed it. Thank you.

File: src/test/java/test_cases/RegisterTests.java (hunk 13 above)
Comment (viethuynh140499) [reply src/test/java/test_cases/RegisterTests.java:26]: I fixed it. Thank you.

File: src/test/java/test_cases/ChangePasswordTest.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    homePage.clickChangePasswordTab();
+
+    String newPassword = DataHelper.getRandomValidPassword();
+    user.setPassword(newPassword);
+
+    changePasswordPage.changePassword(oldPassword, newPassword, newPassword);
+    String actualSuccessMessage = changePasswordPage.getSuccessfullyMessage();
+    String expectedSuccessMessage = "Your password has been updated!";
+
+    Assert.assertEquals(actualSuccessMessage, expectedSuccessMessage, "Change password unsuccessfully");
```
Comment (tanmle) [inline src/test/java/test_cases/ChangePasswordTest.java:39]: you should also check that user is able to login with new password as well.

File: src/main/java/page_objects/RegisterPage.java (hunk 7 above)
Comment (viethuynh140499) [reply src/main/java/page_objects/RegisterPage.java]: I changed By.id

File: src/test/java/test_cases/ChangePasswordTest.java (hunk 14 above)
Comment (viethuynh140499) [reply src/test/java/test_cases/ChangePasswordTest.java:39]: I fixed it , thank you

File: src/test/java/test_cases/RegisterTests.java (hunk 13 above)
Comment (thanhlelgg) [reply src/test/java/test_cases/RegisterTests.java:27]: it's till there?

**Review by fey031 — APPROVED** [review]
LGTM

File: src/test/java/test_cases/RegisterTests.java (hunk 13 above)
Comment (viethuynh140499) [reply src/test/java/test_cases/RegisterTests.java]: Sorry, I fixed it. Thank you

**Pull request merged**